### ISSUE TITLE
New class for returning error objects?

### ISSUE THREAD

**Issue by michalstolarczyk** [pepkit/BiocProject#16]
The `BiocProject` function catches errors and warnings produced by the user-provided data reading function. As we decided before (e.g, see #6) the error messages should be returned instead of the read data in case these occur. 

With recent changes in the `BiocProject` package design the returned object  has to be of class `Annotated` to contain the metadata as the `Project` object.  Therefore to return the error message (and be consistent) we need to use an object that extends the `Annotated` to enclose the error messages. `Annotated` is a virtual class, so we can't use it directly. Here are other options:

```r
> showClass("Annotated")
Virtual Class "Annotated" [package "S4Vectors"]

Slots:
               
Name:  metadata
Class:     list

Known Subclasses: 
Class "Vector", directly
Class "Hits", by class "Vector", distance 2
Class "SelfHits", by class "Hits", distance 3
Class "SortedByQueryHits", by class "Hits", distance 3
Class "SortedByQuerySelfHits", by class "SelfHits", distance 4
Class "Rle", by class "Vector", distance 2
Class "List", by class "Vector", distance 2
Class "SimpleList", by class "List", distance 3
Class "HitsList", by class "SimpleList", distance 4
Class "SelfHitsList", by class "HitsList", distance 5
Class "SortedByQueryHitsList", by class "HitsList", distance 5
Class "SortedByQuerySelfHitsList", by class "SelfHitsList", distance 6
Class "Pairs", by class "Vector", distance 2
Class "FilterRules", by class "SimpleList", distance 4
```

We could use class `List` to store the error message (this is how it's temporarily done now). However, there is one small drawback; the `show` method of the `List` class *does not* display the contents of the `@listData` slot by default: 

```r
> l = List("test")
> l
List of length 1
> l@listData
[[1]]
[1] "test"
```

So I propose we write a class (`AnnotatedError`,`ListError`?) that would extend the `List` class and override the inherited `show` method with one that displays errors nicely.

**Comment by michalstolarczyk** [comment]
So I guess there's no need to implement new `AnnotatedError` class if we allow for returning other objects than `Annotated`. See #19

**Comment by michalstolarczyk** [comment]
Or we can use the `S4Vectors` `List` class in #19 that extends `Annotated` class and stick to the "return `Annotated` only" rule.

**Comment by nsheff** [comment]
```
> metadata(a) = pepr::Project()
Error in .local(x, ..., value) : 
  replacement 'metadata' value must be a list
> metadata(a) = list(pepr::Project())
> a
List of length 3
names(3): a b c
> metadata(a)
[[1]]
PEP project object. Class:  Project
  file:  
  samples:  0

> a
List of length 3
names(3): a b c
> List(1,2,3)
NumericList of length 3
[[1]] 1
[[2]] 2
[[3]] 3
> List(a=1,b=2,c=3)
NumericList of length 3
[["a"]] 1
[["b"]] 2
[["c"]] 3
> a = List(a=1,b=2,c=3)
> a
NumericList of length 3
[["a"]] 1
[["b"]] 2
[["c"]] 3
> metadata(a) = list(pepr::Project())
> a
NumericList of length 3
[["a"]] 1
[["b"]] 2
[["c"]] 3
> 
```

**Comment by michalstolarczyk** [comment]
Resolved with: 811ff784ed990ac7a066b13597a6506150b045f4

the returned object/error does not have to be of class `Annotated`. The `S4Vectors::List` object is returned if this happens. And the metadata slot is populated with the `pepr::Project` object

**Comment by michalstolarczyk** [comment]
In the end, the `BiocProject` function always returns `Annotated`
